src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter.cs

```diff
@@ -24,12 +36,65 @@ public void OnTempDataSaving(ITempDataDictionary tempData)
                     var originalValue = kvp.Value;
 
                     var newValue = property.GetValue(Subject);
-                    if (newValue != null && newValue != originalValue)
+                    if (newValue != null && !newValue.Equals(originalValue))
                     {
                         tempData[Prefix + property.Name] = newValue;
                     }
                 }
             }
         }
+
+        public void OnActionExecuting(ActionExecutingContext context)
```

Comment:
We wouldn't add an async API to `ITempDataDictionary`, it would go on the factory/provider. As far as doing this async goes, I'm not fully convinced that we need it. Separate issue maybe?